chore(storybook): added static storybook build

kristofdegrave opened the pull request:
_[Please make sure you have read the submission guidelines before posting an PR](https://github.com/nrwl/nx/blob/master/CONTRIBUTING.md#submit-pr)_

> _Please make sure that your commit message follows our format._

> Example: `fix(nx): must begin with lowercase`

## Current Behavior (This is the behavior we have today, before the PR is merged)
Its only possible to make a dev server build to run storybook local

## Expected Behavior (This is the new behavior we can expect after the PR is merged)
Added a production build builder to generate a static build that can be published to a webserver

## Issue

kristofdegrave commented:
Ok, I'll have a look at it and see how I can implement it. The only thing I doubt about is the naming. Currently it is `storybook-static`, but maybe we should rename the dev and prod build to something like `storybook:serve` and `storybook:build` to align with the other build and serve tasks?

isaacplmann commented:
You're right, the names should probably be `dev-server` and `build` to align with the other builders, but I'd need to write a migration script to rename the existing builder.  Why don't you change your builder to `build` and set the task name to `storybook-build`.  So the final `workspace.json`/`angular.json` would look like this:

```
"some-project": {
  "architect": {
    "storybook": {
      "builder": "@nrwl/storybook:storybook",
      // etc...
    },
    "storybook-build": {
      "builder": "@nrwl/storybook:build",
      // etc...
    }
  }
}
```

In a separate PR I can work on migrating this:

```
    "storybook": {
      "builder": "@nrwl/storybook:storybook",
      // etc...
    },
```

To this:

```
    "storybook-serve": {
      "builder": "@nrwl/storybook:dev-server",
      // etc...
    },
```

kristofdegrave commented:
I renamed the task as proposed, I also renamed the folders following the convention.

I also added an e2e test that currently only checks if an index.html file is added in the supposed drop folder of the storybook build folder. I need to investigate how I can improve the current e2e test, but I had some issues setting up my environment to run the e2e tests (windows environment) and the e2e tests seem to run extremely slow on the local environment.

isaacplmann commented:
The e2e tests are pretty slow.  But you can run just the storybook e2e test with `yarn e2e storybook`.

rommandz commented:
Hi! I'm interested in this pr for the purposes of my project. 

I've mentioned that the static builder in the current implementation is running the node task after a successful build instead of exiting with zero code from the task.

 I think most users expect from build task that it should terminate after a successful job, or it better if there will be some configuration flag available like:

`watch: boolean`

for more complex flexibility.

What do you think?

isaacplmann commented:
@rommandz , the current implementation isn't a build task, but a server task.  That's why it doesn't terminate when it finishes.  This PR is adding a build task that will create static files on disk for you.

I'd argue against adding a `watch: boolean` flag for two reasons:
1. There are quite a few configuration options that are different between the `serve` mode and the `build` mode.  It would be difficult to explain which options are valid in which mode.  Keeping the builders separate makes this obvious.

2. All the other builders in the `@nrwl` packages have separate builders for `serve` and `build`.

rommandz commented:
@isaacplmann , maybe there was some misunderstanding.

By the current implementation, I meant the implementation of the new builder (for build static files) in current PR.

It's simple and clear why the `storybook` builder (which already available on master) for dev server doesn't terminate.

But builder for static files (`storybook-static` aka `storybook:build`) which added in this PR should terminate after a successful build, isn't it?

If no, I really don't understand how to use it for example in monorepo with storybook project and `@nrwl/affected build.`

isaacplmann commented:
@rommandz yes, sorry, I misunderstood.  The builder in this PR should terminate when it finishes.  We'll make sure not to merge unless it does.

isaacplmann commented:
Running `yarn format` will fix the first build failure.  It looks like the second build failure is because the storybook build process never actually exits.

isaacplmann commented:
I think if you change the `runInstance` function to this, the builder will complete successfully.  Your implementation was starting to do what `from` does but didn't `complete` the new observable.

```
function runInstance(options: StorybookBuilderOptions) {
  return from(buildStaticStandalone({ ...options, ci: true }));
}
```

isaacplmann commented:
I'd also recommend taking these lines:

```
      const storybookOutput = runCLI(`run ${mylib}:storybook-build`);
      //expect(storybookOutput).toContain('info => Preview built');

      checkFilesExist(`dist/storybook/${mylib}/index.html`);
      expect(readFile(`dist/storybook/${mylib}/index.html`)).toContain(
        `<title>Storybook</title>`
      );
```

And putting them at the end of the existing storybook e2e test to save time on test runs.

kristofdegrave commented:
Tnx for the insights I'll try them out.

Running with the nx cli seems to work fine. It's using the angular cli that seems to run into an issue in not completing. I see the files being generated, but the task doesn't complete. I tried debugging it, but this corrupted my pc today :). Seems like WSL isn't that stable yet...

kristofdegrave commented:
The changes you mentioned worked. Sorry for the hugh number of commits, but WSL doesn't seem to be very stable yet, so it was hard to test parts locally.

kristofdegrave commented:
finally a green pipeline :) @isaacplmann

vsavkin commented on scripts/e2e-ci2.sh:
```diff
@@ -5,7 +5,7 @@
 mkdir -p tmp/angular
 mkdir -p tmp/nx
-
+ 
```
remove

vsavkin commented on packages/storybook/src/schematics/configuration/lib-files/.storybook/webpack.config.js__tmpl__:
```diff
@@ -15,5 +15,5 @@ module.exports = async ({ config, mode }) => {
     }
   });<% } %>
-  return config;
+  return config; 
```
remove

vsavkin commented on package.json:
```diff
@@ -8,11 +8,11 @@
     "e2e-rerun": "./scripts/e2e-rerun.sh",
     "e2e-ci1": "./scripts/e2e-ci1.sh",
-    "e2e-ci2": "./scripts/e2e-ci2.sh",
+    "e2e-ci2": "./scripts/e2e-ci2.sh", 
```
remove

vsavkin commented on package.json:
```diff
@@ -8,11 +8,11 @@
     "build": "./scripts/package.sh --local",
     "commit": "git-cz",
-    "checkcommit": "node ./scripts/commit-lint.js",
+    "checkcommit": "node ./scripts/commit-lint.js", 
```
remove

isaacplmann commented:
All the PRs are failing the node e2e test.  #2163 looks like it will fix this.  We'll rebase once that is merged in.

kristofdegrave commented:
Looking forward to it :)

FrozenPandaz commented on packages/storybook/src/schematics/configuration/configuration.ts:
```diff
@@ -106,5 +107,24 @@ function addStorybookTask(projectName: string, uiFramework: string): Rule {
         }
       }
     });
+    projectConfig.targets.set('storybook-build', {
```
nitpick:

Can you please set the target as `build-storybook` so it's a verb like other target names?

kristofdegrave replied:
ofc, name was already changed to `storybook-build` to get compliant

FrozenPandaz commented on packages/storybook/src/builders/storybook-build/storybook-build.impl.spec.ts:
```diff
@@ -0,0 +1,5 @@
+describe('storybook builer', () => {
+  it('should have a test', () => {
```
Are there any more unit tests that would be useful here?

kristofdegrave replied:
I can have a look at this, but my feeling is that this will be equal to the e2e test. If you have any suggestions on how to tackle this, please let me know.

FrozenPandaz replied:
I think testing the webpack config which gets used here is useful. Maybe look at the way the web and node build builders are tested :wink: .

kristofdegrave replied:
@FrozenPandaz I have tried to work an a similar way like the web builder spec, but I was able to get a successful run. It start running, but on the part it starts compiling it stats to go wrong
```
 clean outputDir..
 => Copying prebuild dll's..
 => Building manager..
 => Loading manager config..
 => Loading presets
 => Compiling manager..

ERR! => Failed to build the manager
ERR!   TypeError: Cannot read property '_location' of null
ERR!   
ERR!   - Window.js:223 Window.get location [as location]
ERR!     [nx]/[jest-environment-jsdom]/[jsdom]/lib/jsdom/browser/Window.js:223:79
ERR!   
ERR!   - lodash.js:2392 arrayLikeKeys
ERR!     [nx]/[lodash]/lodash.js:2392:16
ERR!   
ERR!   - lodash.js:13334 keysIn
ERR!     [nx]/[lodash]/lodash.js:13334:36
ERR!   
ERR!   - lodash.js:12639 
ERR!     [nx]/[lodash]/lodash.js:12639:26
ERR!   
ERR!   - lodash.js:4853 Function.<anonymous>
ERR!     [nx]/[lodash]/lodash.js:4853:13
ERR!   
ERR!   - lodash.js:475 apply
ERR!     [nx]/[lodash]/lodash.js:475:27
ERR!   
ERR!   - lodash.js:6563 Function.<anonymous>
ERR!     [nx]/[lodash]/lodash.js:6563:16
ERR!   
ERR!   - index.js:336 HtmlWebpackPlugin.evaluateCompilationResult
ERR!     [nx]/[core]/[html-webpack-plugin]/index.js:336:42
ERR!   
ERR!   - index.js:258 
ERR!     [nx]/[core]/[html-webpack-plugin]/index.js:258:25
ERR!   
ERR!   - async Promise.all
ERR!   
ERR! 
(node:2071) UnhandledPromiseRejectionWarning: Error: 
    at Env.it (/mnt/d/GIT/nx/node_modules/jest-jasmine2/build/jasmineAsyncInstall.js:91:24)
    at it (/mnt/d/GIT/nx/node_modules/jest-jasmine2/build/jasmine/jasmineLight.js:93:21)
    at Suite.<anonymous> (/mnt/d/GIT/nx/build/packages/storybook/src/builders/build-storybook/build-storybook.impl.spec.js:20:5)
    at addSpecsToSuite (/mnt/d/GIT/nx/node_modules/jest-jasmine2/build/jasmine/Env.js:496:51)
    at Env.describe (/mnt/d/GIT/nx/node_modules/jest-jasmine2/build/jasmine/Env.js:466:11)
    at describe (/mnt/d/GIT/nx/node_modules/jest-jasmine2/build/jasmine/jasmineLight.js:81:18)
    at Object.<anonymous> (/mnt/d/GIT/nx/build/packages/storybook/src/builders/build-storybook/build-storybook.impl.spec.js:15:1)
    at Runtime._execModule (/mnt/d/GIT/nx/node_modules/jest-runtime/build/index.js:867:68)
    at Runtime._loadModule (/mnt/d/GIT/nx/node_modules/jest-runtime/build/index.js:577:12)
    at Runtime.requireModule (/mnt/d/GIT/nx/node_modules/jest-runtime/build/index.js:433:10)
(node:2071) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:2071) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
error Command failed with exit code 1.
```

isaacplmann replied:
@kristofdegrave I can take a look at the unit test code, but I don't think you pushed it up

kristofdegrave replied:
@isaacplmann tnx! I've pushed the changes now.

isaacplmann replied:
In `storybook/src/utils/testing.ts`

```
import { createEmptyWorkspace } from '@nrwl/workspace/src/utils/testing-utils';
```

I found this issue by running the test in debug mode:

```
node --inspect-brk ./node_modules/jest/bin/jest.js build/packages/storybook/src/builders/build-storybook/build-storybook.impl.spec.js
```

kristofdegrave replied:
@isaacplmann tnx for the hint! Unit test now seem to work :)

kristofdegrave replied:
@isaacplmann it seems the tests are running local, but it keeps giving an issue on the buildserver

`Unhandled rejection TypeError: setTimeout(...).unref is not a function`

isaacplmann commented:
@kristofdegrave I was able to run all the tests with `yarn test` only after I removed the `should run successfull` test.  Before that, the jest process ran out of memory.  I think it's safe to remove that particular test, since that functionality is being covered by the e2e test.

isaacplmann commented:
Also, the other three tests could be probably be consolidated into a single test with three expectations.

kristofdegrave commented:
@isaacplmann tnx for the feedback.

Changes are applied an I have requested reviews of @FrozenPandaz  and @vsavkin again.

FrozenPandaz commented on packages/storybook/src/utils/testing.ts:
```diff
@@ -92,3 +107,110 @@ export class TestButtonComponent implements OnInit {
+
+  addTeardown(teardown: () => Promise<void> | void) {}
+
+  async getProjectMetadata(
```
Sorry for the inconvenience:

Can you please merge in the latest master and remove this method.

Your branch has changes (angular-devkit 9.0.0) which has been reverted for now.

This method is not available yet. I can handle adding it back when it is.
